assembl/scripts/ini_files.py

```diff
@@ -144,72 +166,392 @@ def main():
+                changed = True
+    if changed:
+        with open(random_file, 'w') as f:
+            base.write(f)
+    return base
+
+
+def rc_to_ini(rc_info, default_section=SECTION):
+    """Convert a .rc file to a ConfigParser (.ini-like object)
+
+    Items are assumed to be in app:assembl section,
+        unless prefixed by "{section}__" .
+    Keys prefixed with an underscore are not passed on.
+    Keys prefixed with a star are put in the global (DEFAULT) section.
+    Value of '__delete_key__' is eliminated if existing.
+    """
+    p = Parser()
+    for key, val in rc_info.iteritems():
+        if not key or key.startswith('_'):
+            continue
+        if key[0] == '*':
+            # Global keys
+            section = "DEFAULT"
+            key = key[1:]
+        elif '__' in key:
+            section, key = key.split('__', 1)
+        else:
+            section = default_section
+        ensureSection(p, section)
+        if val == '__delete_key__':
+            # Allow to remove a variable from rc
+            # so we can fall back to underlying ini
+            p.remove_option(section, key)
+        else:
+            p.set(section, key, val)
+    return p
+
+
+def iniconfig_to_rc(parser, dest=None, extends=None, target_dir=None):
+    """Convert a ConfigParser to a .rc file.
+
+    That file is written to dest, or to a returned file-like object
+    extends specifies an extended ini file."""
+    from cStringIO import StringIO
+    if dest is None:
+        dest = StringIO()
+    if extends:
+        if target_dir is not None:
+            target_dir = os.path.abspath(target_dir)
+            extends_dir = os.path.abspath(os.path.dirname(extends))
+            extends = os.path.join(os.path.relpath(extends_dir, target_dir),
+                                   os.path.basename(extends))
+        dest.write("_extends = %s\n" % (extends,))
+    for section in parser.sections():
+        if section == SECTION:
+            prefix = ''
+        else:
+            prefix = section + '__'
+        for key, value in parser.items(section):
+            if '\n' in value:
+                log.warning("avoid multiline values in RC: %s=%s", key, value)
+                value = ' '.join(value.split('\n'))
+            dest.write("%s%s = %s\n" % (prefix, key, value))
+    if hasattr(dest, 'seek'):
+        dest.seek(0)
+    return dest
+
+
+def extract_saml_info(rc_info):
+    """Extract SAML variables from the state"""
+    saml_info = {k[5:]: v for (k, v) in rc_info.iteritems()
+                 if k.startswith('saml_')}
+    saml_info['cn'] = rc_info['public_hostname']
+    return saml_info
+
+
+def diff_ini(first, second, diff=None, existing_only=False):
+    """Diff ini files
+
+    Generate a parser with any value in the second that is different in the first.
+    Returns a ConfigParser.
+    Takes interpolation into account. Does not include values that disappeared."""
+    from ConfigParser import _Chainmap
```

Comment:
It's ConfigParser in python2, configparser in python 3. Which docs are your reading?